server/src/main/java/org/elasticsearch/action/update/TransportUpdateAction.java

```diff
@@ -333,4 +343,58 @@ private void handleUpdateFailureWithRetry(
         }
         listener.onFailure(cause instanceof Exception ? (Exception) cause : new NotSerializableExceptionWrapper(cause));
     }
+
+    private static UpdateHelper.Result deleteInferenceResults(
+        UpdateRequest updateRequest,
+        UpdateHelper.Result result,
+        IndexMetadata indexMetadata,
+        MappingLookup mappingLookup
+    ) {
+        if (result.getResponseResult() != DocWriteResponse.Result.UPDATED) {
+            return result;
+        }
+
+        Map<String, InferenceFieldMetadata> inferenceFields = indexMetadata.getInferenceFields();
+        if (inferenceFields.isEmpty()) {
+            return result;
+        }
+
+        Map<String, Object> updateRequestSource = updateRequest.doc().sourceAsMap();
+        Map<String, Object> updatedSource = result.updatedSourceAsMap();
+        boolean updatedSourceModified = false;
+        for (var entry : inferenceFields.entrySet()) {
+            String inferenceFieldName = entry.getKey();
+            Mapper mapper = mappingLookup.getMapper(inferenceFieldName);
+
+            if (mapper instanceof InferenceFieldMapper inferenceFieldMapper) {
+                String[] sourceFields = entry.getValue().getSourceFields();
+                for (String sourceField : sourceFields) {
+                    if (sourceField.equals(inferenceFieldName) == false && updateRequestSource.containsKey(sourceField)) {
```

Comment:
We should use `XContentMapValues.extractValue` to check if the `sourceField` is present. We need to handle dots in field names and object fields.

Reply:
Good catch! And we need a test for this :)